### PR TITLE
fix(mock_terraform): mock terraform for testing

### PR DESCRIPTION
**Pull request by thanh-nguyen-dang**
r @philloooo: mock terraform test, create_instances is ready

### CODE REVIEW

**Comment by philloooo** [comment]
add a decorator for those instances endpoints which check if there exists a vpc and subnet for this user?
right now it just 500 if there is not 'subnet_id' in additional_info

File: cloud_middleware/cloud_driver/terraform.py [hunk 1]
```diff
@@ -69,5 +69,8 @@ class VPCOutput():
     def __init__(self, vpc_id, vpc_name, security_group_ids, subnet_id):
         self.vpc_id = vpc_id
         self.vpc_name = vpc_name
+        # self.security_group_ids = []
+        # for s in security_group_ids:
+        #     self.security_group_ids.append({"id": s})
```
Comment (philloooo) [inline cloud_middleware/cloud_driver/terraform.py]: cleanup?

File: middleware_test/api_test/test_aws.py [hunk 2]
```diff
@@ -11,11 +11,13 @@
...
-        'imageId': 'ami-6869aa05',
+        'image_id': 'ami-6869aa05',
         'count': 3,
-        'instanceType': 't2.small',
-        'externalId': 'abc',
+        'instance_type': 't2.small',
+        'external_id': 'abc',
         'name': 'test_instance'
     }
     post_msg = json.dumps(json_msg)
```
Comment (philloooo) [inline middleware_test/api_test/test_aws.py]: verify that `client.post("/aws/v0/instances", data=post_msg)` returns the instance with correct name and key_name?

File: cloud_middleware/api/aws.py [hunk 3]
```diff
@@ -198,9 +197,23 @@ def get_vpc():
 
     d_results = {k: v for k, v in user_info.items()
                  if (k in ["vpc_id", "vpc_name", "security_group_ids", "subnet_id"])}
-
     return jsonify(d_results)
 
+
+@rest.post('/vpc_test')
```
Comment (philloooo) [inline cloud_middleware/api/aws.py]: why do we have a `test` endpoint for production api?

File: cloud_middleware/cloud_driver/aws_driver.py [hunk 4]
```diff
@@ -60,22 +60,32 @@ def setup_vpc(self, user_info):
...
+                         user_data=None, placement=None, kernel_id=None, ramdisk_id=None,
+                         block_device_mappings=None, monitoring=None, disable_api_termination=False,
                          instance_initiated_shutdown_behavior=None, private_ip_address=None,
                          client_token=None, additional_info=None, network_interfaces=None,
-                         iam_instance_profile=None, ebs_optimized=False,
+                         iam_instance_profile=None, ebs_optimized=False, name=None,
                          **kwargs):
         ec2, credentials = self.login_with_user_info(user_info)
         # Launch vms #
         self.logger.debug("user_info: %s; image_id: %s; instance_type: %s",
```
Comment (philloooo) [inline cloud_middleware/cloud_driver/aws_driver.py]: the image_id is passed from user according to your code, but we should only allow image id that we approved, so we should keep a list of allowed image id and a default image id on the api side

File: cloud_middleware/api/aws.py [hunk 5]
```diff
@@ -125,10 +125,8 @@ def delete_instances(instance_ids):
 def launch_instances(data):
     # logger = rest.cloud_driver.logger
     server_data = json.loads(data)
-    # logger.debug(server_data)
     user_info = get_current_user_info()
-    instances = rest.cloud_driver.launch_instances(user_info, server_data["imageId"],
-                                                   server_data["instanceType"],
+    instances = rest.cloud_driver.launch_instances(user_info,
                                                    max_count=int(server_data["count"]),
                                                    min_count=int(server_data["count"]),
```
Comment (philloooo) [inline cloud_middleware/api/aws.py]: default count to 1?
`server_data.get('count', 1)`

Comment (thanh-nguyen-dang) [inline cloud_middleware/api/aws.py]: 👍

File: cloud_middleware/cloud_driver/terraform.py (hunk 1 above)
Comment (thanh-nguyen-dang) [inline cloud_middleware/cloud_driver/terraform.py]: added a decorator to check the existing of a vpc and subnet

File: middleware_test/api_test/test_aws.py (hunk 2 above)
Comment (thanh-nguyen-dang) [inline middleware_test/api_test/test_aws.py]: fixed

File: cloud_middleware/cloud_driver/aws_driver.py (hunk 4 above)
Comment (thanh-nguyen-dang) [inline cloud_middleware/cloud_driver/aws_driver.py]: fixed

File: cloud_middleware/cloud_driver/aws_driver.py [hunk 6]
```diff
@@ -174,3 +174,13 @@ def create_key_pair(self, user_info, name):
     def list_key_pairs(self, user_info):
         ec2, credentials = self.login_with_user_info(user_info)
         return ec2.meta.client.describe_key_pairs()
+
+    def set_instance_to_state(self, instance_id, to_state, **kwargs):
+        user_info = kwargs.get('user_info', None)
+        ec2, credentials = self.login_with_user_info(user_info)
+        if to_state == 'stop':
+            return ec2.meta.client.stop_instances(InstanceIds=[instance_id], Force=True)
+        if to_state == 'reboot':
```
Comment (philloooo) [inline cloud_middleware/cloud_driver/aws_driver.py]: `elif` to short circuit?
Also 
```else:
  raise UserError("Invalid transition state")
```
?

Comment (thanh-nguyen-dang) [inline cloud_middleware/cloud_driver/aws_driver.py]: added error message.

File: cloud_middleware/api/aws.py (hunk 3 above)
Comment (thanh-nguyen-dang) [inline cloud_middleware/api/aws.py]: remove it. fixed with a decorator

File: middleware_test/api_test/test_aws.py [hunk 7]
```diff
@@ -69,7 +70,15 @@ def test_create_delete_instance(client, userapi_client):
     headers = {'content-type': 'application/json'}
     r = userapi_client.post("/aws/role_arn", data=post_msg, headers=headers)
     list_first_ids = create_instances_and_verify(client)
-    delete_created_instace(client, list_first_ids)
+
+    # response = delete_created_instances(client, list_first_ids)
+    response = set_instance_state(client, list_first_ids[0])
```
Comment (philloooo) [inline middleware_test/api_test/test_aws.py]: can you verify the content of the result?

File: cloud_middleware/api/aws.py [hunk 8]
```diff
@@ -191,39 +219,22 @@ def generator(sp):
...
+@login_required
+@vpc_required
+def change_instance_state(instance_id, data, user_info=None):
+    capp.m_logger.debug("instance_id: {0}".format(instance_id))
+    capp.m_logger.debug("data: {0}".format(data))
+    request_data = json.loads(data)
+    to_state = request_data['to_state']
+    result = rest.cloud_driver.set_instance_to_state(instance_id, to_state, user_info=user_info)
+    result = {k: v for k, v in result.items()
+              if (k in ['StoppingInstances', 'Rebooting'])}
```
Comment (philloooo) [inline cloud_middleware/api/aws.py]: Can you provide a generic result? The expected response would be same as the 
`POST /instances/<instance_id>` endpoint, which return the updated instance information

Comment (thanh-nguyen-dang) [inline cloud_middleware/api/aws.py]: use output format: {previous_state: prev, current_state:current}
if you feel good, I will merge this pull request

**Pull request merged**